### PR TITLE
Use `invoke_result` if available for supporting C++20

### LINKED ISSUE #5082: [compile error] C++20 removed error std::result_of used in <pcl/common/common.h>
**Describe the error**
Using the VCPKG port of PCL with Version: 1.12.0

C++20 removed the `std::result_of` call (see https://en.cppreference.com/w/cpp/types/result_of) which results in compile errors when including `<pcl/common/common.h>`.

```c++
template<typename IteratorT, typename Functor> inline auto
computeMedian (IteratorT begin, IteratorT end, Functor f) noexcept -> typename std::result_of<Functor(decltype(*begin))>::type
```

The exact compile error is as follows:
`pcl/common/common.h(285): error C2653: 'result_of': is not a class or namespace name`

**To Reproduce**
Install PCL via vcpkg and in a MSVC project using /std:c++20 include `<pcl/common/common.h>`.

or

Just enable a C++20 compliant compiler when including `<pcl/common/common.h>`

**Your Environment (please complete the following information):**

 - OS: Windows 10
 - Compiler: Visual Studio 2019 16.11.7
 - PCL Version 1.12.0
 - PCL Type: Compiled via VCPKG with /arch:AVX

**Possible Solution**
When detecting a compiler with C++17 or greater use `std::invoke_result` instead (https://en.cppreference.com/w/cpp/types/result_of).

### PR DESCRIPTION
**Pull request by kunaltyagi**
Potentially, closes #5082

### CODE REVIEW

**Comment by kunaltyagi** [comment]
@cbrammere3d PTAL. If this solves your issue, please let us know

**Comment by cbrammere3d** [comment]
Looks good to me. Thanks for the quick PR

**Review by larshg — DISMISSED** [review]

File: common/include/pcl/common/common.h [hunk 1]
```diff
@@ -282,7 +282,12 @@ namespace pcl
     * \ingroup common
     */
   template<typename IteratorT, typename Functor> inline auto
-  computeMedian (IteratorT begin, IteratorT end, Functor f) noexcept -> typename std::result_of<Functor(decltype(*begin))>::type
+  computeMedian (IteratorT begin, IteratorT end, Functor f) noexcept ->
+  #if __cpp_lib_is_invocable
+  std::invoke_result_t<Functor(decltype(*begin))>
```
Comment (mvieth) [inline common/include/pcl/common/common.h:287]:
```suggestion
  std::invoke_result_t<Functor, decltype(*begin)>
```
See https://en.cppreference.com/w/cpp/types/result_of

**Comment by kunaltyagi** [comment]
I thought when OP says it works, it worked 😆

**Comment by kunaltyagi** [comment]
Though this has made me think we should move 1 of the CI to C++17 at least. Which one would you prefer?

**Review by mvieth — APPROVED** [review]
PR looks good now.
But I am not so sure if we should change one of the CIs to another C++ standard, since PCL is still C++14 and we should focus on that. How about creating an additional CI that runs e.g. weekly (not for every PR) and covers less common build configurations, e.g. C++17 and C++20, other values for `CMAKE_BUILD_TYPE` etc

**Comment by kunaltyagi** [comment]
Weekly CI is a good idea. I just hate that they don't have good visibility (but we can create a badge for README)

**Review by larshg — APPROVED** [review]

**Pull request merged**